### PR TITLE
Implement 2-cohomology and module computations for arbitrary finite groups, not just solvable ones, via `TwoCohomologyGeneric`

### PR DESCRIPTION
**Pull request by hulpke**
Uses rewriting system since pc presentation will not always be available.
This is both code for the cohomology group, and for constructing the extension corresponding to a particular cocycle. Group is constructed by default as Fp group, but there is an option to have it compute a faithful permutation representation.

Also includes improved routines for `IrreducibleModules` for the non-pc case once such cohomology routines exist, and action of compatible pairs on the cohomology group.

There is a test in `testextra` that, as application, constructs perfect groups anew. (This could easily go to larger orders, if desired, at the moment two quick tests are in.)

### CODE REVIEW

File: doc/ref/grppc.xml
Comment (fingolfin) [inline doc/ref/grppc.xml]:
```suggestion
Section <Ref Sect="2-Cohomology"/> describes operations and methods for
```

Comment (hulpke) [reply doc/ref/grppc.xml]: Thanks.

**Comment by wilfwilson** [comment]
@hulpke I noticed you’ve removed the WIP from the title and the ‘do not review’ label. Just to check, for sure: does this mean that you’d be happy for the PR to be reviewed, and hopefully merged? If so, I’d suggest removing the ‘do not merge’ label because I get the impression it might put people off from reviewing. Sorry I’ve not had chance to look at your PR further yet.

**Comment by alex-konovalov** [comment]
Travis reports two manual example diffs:
```
########> Diff in /home/travis/build/gap-system/gap/doc/ref/../../lib/twocohom\
.gd:273 (/home/travis/build/gap-system/gap/tst/testmanuals/ref-chapter039.tst:
992)
# Input is:
g1:=Image(IsomorphismPermGroup(p));
# Expected output:
<permutation group of size 38880 with 9 generators>
# But found:
<permutation group with 9 generators>
########
########> Diff in /home/travis/build/gap-system/gap/doc/ref/../../lib/twocohom\
.gd:273 (/home/travis/build/gap-system/gap/tst/testmanuals/ref-chapter039.tst:
999)
# Input is:
g2:=Image(IsomorphismPermGroup(p));
# Expected output:
<permutation group of size 38880 with 9 generators>
# But found:
<permutation group with 9 generators>
########
```

**Comment by hulpke** [comment]
@wilfwilson 
At this stage I am happy for comments about potential problems, but there is still some code (basic functionality for compatible pairs and action on the cohomology group) I am still working on and want to add before merging. (There also is a manual bug and a stupid typo in the perfect group construction example that need fixing.)
As there will still be code added (as soon as I fix a bug..), it might be best to wait with formal reviewing.

**Comment by hulpke** [comment]
@wilfwilson OK, now the promised changes are in, and the holds are removed.
Since there is no need to backport, I'm not sure whether there is a need to squash into one commit, but could do so if desired.

**Review by fingolfin — CHANGES_REQUESTED** [review]
Very nice, thank you! Overall, this is of course a very welcome improvement!

Unfortunately I did not yet have time to look at everything (with easter and everything), but I left a few rather minor remarks. I would like to have a closer look at a few more things, I'll try to get this done this week or early next week.

Of course it would be nice to have this in GAP 4.11 -- @alex-konovalov is in charge of that.

File: lib/grpreps.gi
Comment (hulpke) [reply lib/grpreps.gi]: Will change so (but it anyhow is a fallback method, permutation groups use Burnside/Brauer method below).

File: lib/gpprmsya.gi
Comment (hulpke) [reply lib/gpprmsya.gi]: Done.

File: lib/gpprmsya.gi
Comment (hulpke) [reply lib/gpprmsya.gi]: Changed

File: lib/gpprmsya.gi
Comment (hulpke) [reply lib/gpprmsya.gi]: Done

File: lib/gpprmsya.gi
Comment (hulpke) [reply lib/gpprmsya.gi]: I have replaced tabs.

File: lib/gpprmsya.gi
Comment (hulpke) [reply lib/gpprmsya.gi]: No, `gens` are free generators, while `F` now is fp.

File: lib/twocohom.gi
Comment (fingolfin) [inline lib/twocohom.gi]: Also not covered by `testinstall`.

File: lib/twocohom.gi
Comment (fingolfin) [inline lib/twocohom.gi]: Also not covered by `testinstall`.

File: lib/twocohom.gi
Comment (fingolfin) [inline lib/twocohom.gi]: Also not covered by `testinstall`.

File: lib/twocohom.gi
Comment (fingolfin) [inline lib/twocohom.gi]: Also not covered by `testinstall`.

File: lib/grppcext.gi
Comment (alex-konovalov) [reply lib/grppcext.gi]: I could run the test of this PR in Jenkins, to exercise testextra. However, it does not pass cleanly at the moment: #3408

Comment (hulpke) [reply lib/grppcext.gi]: @fingolfin I have added smaller a test to `testinstall' that checks for action of compatible pairs (thus testing most code). Is seems to run through (the current test failure is due to a DNS resolving problem).

Comment (fingolfin) [reply lib/grppcext.gi]: Thanks. I'll go over the coverage once again.

File: lib/grppcext.gi
Comment (fingolfin) [inline lib/grppcext.gi]: This loop is still not covered.

File: lib/grppcext.gi
Comment (fingolfin) [inline lib/grppcext.gi]: Still not covered.

File: lib/grppcext.gi
Comment (fingolfin) [reply lib/grppcext.gi]: Still not covered.

File: lib/grppcext.gi
Comment (fingolfin) [reply lib/grppcext.gi]: Still not covered.

File: lib/grpreps.gi
Comment (fingolfin) [inline lib/grpreps.gi]: Still not covered.

File: lib/grpreps.gi
Comment (fingolfin) [reply lib/grpreps.gi]: Still not covered.

File: lib/grpreps.gi
Comment (fingolfin) [reply lib/grpreps.gi]: Still not covered.

File: lib/twocohom.gi
Comment (fingolfin) [inline lib/twocohom.gi]: This is one of only two lines in this function which is no covered. Very nice!

File: lib/twocohom.gi
Comment (fingolfin) [inline lib/twocohom.gi]: (This is the other uncovered line).

File: lib/twocohom.gi
Comment (fingolfin) [reply lib/twocohom.gi]: This is now covered, thanks.

File: lib/twocohom.gi
Comment (fingolfin) [reply lib/twocohom.gi]: Still not covered.

File: lib/twocohom.gi
Comment (fingolfin) [reply lib/twocohom.gi]: This is now covered, nice!

**Review by fingolfin — APPROVED** [review]
Thank you for the added `testinstall` test. There are still about 300 new lines of code which are not covered by a test in `testinstall`, and the new code only has a coverage rate of ~76% as a result. I'd prefer if it was higher (not because high coverage by itself is a magic goal, but rather because it would increase my confidence in all those special cases, and also in our ability to keep this code from breaking in the future).

Anyway, I don't think we need to delay this further because of that. So let's merge it soon. (I'll wait for Travis to finish with its current truck loads of builds before merging it).

File: lib/twocohom.gi
Comment (hulpke) [reply lib/twocohom.gi]: Yes, this is as a hook for future extensions (RWS for simple groups).

**Pull request merged**
